src/main/java/org/wildfly/security/auth/server/ServerAuthenticationContext.java

```diff
@@ -855,6 +856,22 @@ private void handleOne(final Callback[] callbacks, final int idx) throws IOExcep
                     if (authenticationID != null) {
                         // always re-set the authentication name to ensure it hasn't changed.
                         setAuthenticationName(authenticationID);
+                    } else {
+                        // This is a special case to support scenarios where the identity was already established by some
+                        // external method (e.g.: EXTERNAL SASL and TLS) where only authorization is necessary. We delay authentication
+                        // until we receive an authorization request.
+                        // In the future, we may want to support external methods other than TLS peer authentication
+                        State state = stateRef.get();
+                        if (!(state instanceof NameAssignedState || state instanceof AuthorizedState)) {
+                            if (peerCerts != null) {
+                                log.tracef("Authentication ID is null but SSL peer certificates are available. Trying to authenticate peer");
+                                if (!verifyEvidence(new X509PeerCertificateChainEvidence(peerCerts))) {
```

Comment:
I think we can skip the boolean, if verify does not work then call authorize later - that way it becomes an authorization check for anonymous running as the requested identity.